libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1564,18 +1564,75 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 				setLValueToStorageItem(_indexAccess);
 			break;
 		case DataLocation::Memory:
+			ArrayUtils(m_context).accessIndex(arrayType);
 			setLValue<MemoryItem>(_indexAccess, *_indexAccess.annotation().type, !arrayType.isByteArray());
 			break;
 		case DataLocation::CallData:
-			//@todo if we implement this, the value in calldata has to be added to the base offset
-			solUnimplementedAssert(!arrayType.baseType()->isDynamicallySized(), "Nested arrays not yet implemented.");
-			if (arrayType.baseType()->isValueType())
-				CompilerUtils(m_context).loadFromMemoryDynamic(
-					*arrayType.baseType(),
-					true,
-					!arrayType.isByteArray(),
-					false
-				);
+			if (arrayType.baseType()->isDynamicallyEncoded())
+			{
+				// stack layout: <base_ref> <length> <index>
+				ArrayUtils(m_context).accessIndex(arrayType, true, true);
+				// stack layout: <base_ref> <ptr_to_tail>
+
+				unsigned int baseEncodedSize = arrayType.baseType()->calldataEncodedSize();
+				solAssert(baseEncodedSize > 1, "");
+
+				// returns the absolute offset of the accessed element in "base_ref"
+				m_context.appendInlineAssembly(Whiskers(R"({
+					let rel_offset_of_tail := calldataload(ptr_to_tail)
+					if iszero(slt(rel_offset_of_tail, sub(sub(calldatasize(), base_ref), sub(<neededLength>, 1)))) { revert(0, 0) }
```

Comment:
Does it make sense to introduce some more variables here for this complex expression?

Reply:
Why `baseEncodedSize - 1`?

Reply:
(to turn "not slt", i.e. `>=` into `>`)